components/Banner/examples/supplemental/MDCBannerView.m

```diff
@@ -54,124 +69,312 @@ - (instancetype)initWithCoder:(NSCoder *)aDecoder {
-    self.iconImageViewContainer = nil;
-  }
-  [self setNeedsLayout];
-}
+// TODO: KVO trailingButton.hidden and call `setNeedsUpdateConstraints:`.
 
-- (void)setNumberOfButtons:(NSUInteger)numberOfButtons {
-  NSAssert(numberOfButtons <= kNumberOfButtonsLimit,
-           @"%@ class doesn't support more than %lu buttons", NSStringFromClass([self class]),
-           (unsigned long)kNumberOfButtonsLimit);
-  if (numberOfButtons != self.buttons.count) {
-    NSMutableArray *mutableButtons = [self.buttons mutableCopy];
-    if (numberOfButtons > self.buttons.count) {
-      for (NSUInteger index = self.buttons.count; index < numberOfButtons; ++index) {
-        MDCButton *button = [[MDCButton alloc] init];
-        [mutableButtons addObject:button];
-        [self addSubview:button];
-      }
-    } else if (numberOfButtons < self.buttons.count) {
-      for (NSUInteger index = numberOfButtons; index < self.buttons.count; ++index) {
-        MDCButton *lastObject = [mutableButtons lastObject];
-        [lastObject removeFromSuperview];
-        [mutableButtons removeLastObject];
+#pragma mark - UIView overrides
+
+- (CGSize)sizeThatFits:(CGSize)size {
+  MDCBannerViewLayoutStyle style = [self styleForSizeToFit:size];
+  CGFloat frameHeight = 0.0f;
+  switch (style) {
+    case MDCBannerViewLayoutSingleLineStyle: {
+      frameHeight += kTopPaddingSmall + kBottomPadding;
+      [self.textLabel sizeToFit];
+      [self.leadingButton sizeToFit];
+      CGFloat maximumHeight =
+          MAX(CGRectGetHeight(self.textLabel.frame), CGRectGetHeight(self.leadingButton.frame));
+      if (self.isImageSet) {
+        maximumHeight = MAX(kImageViewSideLength, maximumHeight);
       }
+      frameHeight += maximumHeight;
+      break;
+    }
+    case MDCBannerViewLayoutMultiLineAlignedButtonStyle: {
+      frameHeight += kTopPaddingLarge + kBottomPadding;
+      frameHeight += [self getFrameHeightOfImageViewAndTextLabelWithSizeToFit:size];
+      [self.leadingButton sizeToFit];
+      [self.trailingButton sizeToFit];
+      frameHeight += MAX(CGRectGetHeight(self.leadingButton.frame),
+                         CGRectGetHeight(self.trailingButton.frame));
+      break;
+    }
+    case MDCBannerViewLayoutMultiLineStackedButtonStyle: {
+      frameHeight += kTopPaddingLarge + kBottomPadding;
+      frameHeight += [self getFrameHeightOfImageViewAndTextLabelWithSizeToFit:size];
+      [self.leadingButton sizeToFit];
+      [self.trailingButton sizeToFit];
+      frameHeight += CGRectGetHeight(self.leadingButton.frame) +
+                     CGRectGetHeight(self.trailingButton.frame) + kButtonVerticalIntervalSpace;
+      break;
     }
-    self.buttons = [mutableButtons copy];
   }
+  return CGSizeMake(size.width, frameHeight);
 }
 
-- (NSUInteger)numberOfButtons {
-  return self.buttons.count;
+- (void)layoutSubviews {
+  [super layoutSubviews];
+
+  if (![self isImageSet]) {
+    self.imageView.hidden = YES;
```

Comment:
It is odd to turn this on when you layoutSubviews because it overrides what the client might have set. 

When will we this ever change the hidden property? The method call in the if clause seems to be checking the hidden property already.

Reply:
Yea. This was one part I felt unclean and odd too. The reason why I wanted to do was UIButton set `hidden` when `setImage:nil..` is called. So I want to keep the same behavior. 

I lean more towards using a KVO on `imageView.image`. Addressed in 070341d.